[Xtensa] Implement Xtensa Region Protection Option and several other small Options.

andreisfr opened the pull request:
Implement support of the Xtensa Region Protection, Extended L32R, Data Cache, Relocatable Vector and MISC Special Registers Options.

andreisfr committed:
[Xtensa] Implement Xtensa Region Protection Option and several other small options.

Implement support of the Xtensa Region Protection, Extended L32R,
Data Cache, Relocatable Vector and MISC Special Registers Options.

arsenm commented on llvm/lib/Target/Xtensa/Disassembler/XtensaDisassembler.cpp:
```diff
@@ -114,9 +114,13 @@ static DecodeStatus DecodeMR23RegisterClass(MCInst &Inst, uint64_t RegNo,
-    Xtensa::SAR, 3,  Xtensa::ACCLO,      16, Xtensa::ACCHI,       17,
-    Xtensa::M0,  32, Xtensa::M1,         33, Xtensa::M2,          34,
-    Xtensa::M3,  35, Xtensa::WINDOWBASE, 72, Xtensa::WINDOWSTART, 73};
+    Xtensa::LBEG,    0,   Xtensa::LEND,       1,   Xtensa::LCOUNT,      2,
```
Same comment as the other patch, this is an MCPHysReg + some other number?

andreisfr replied:
I implemented special struct with register name and number fields to make search in SR register table simpler.

arsenm commented on llvm/test/MC/Xtensa/dcache.s:
```diff
@@ -0,0 +1,17 @@
+# RUN: llvm-mc %s -triple=xtensa -show-encoding --mattr=+dcache \
+# RUN:     | FileCheck -check-prefixes=CHECK,CHECK-INST %s
+
+.align	4
```
Do these need mirrored disassembler tests?

andreisfr replied:
Thank you very much for comment! I added disassembler tests for the new  Xtensa Options. Also I found that that we probably should add disassembler tests for Xtensa Options which are already implemented, so I created tests for these Options too.

andreisfr committed:
[Xtensa] Fix disassembler and add tests for Xtensa Options.

Fix disassembling Mac16 Xtensa Option registers and specal registers. Add disasembler tests
for Dcache, ExtendedL32R, Miscsr, Region Protectiom, Relocatable Vecotr, Clamps, Loop, Minmax,
Sext, Nsa, Mul and Div Xtensa Options.

andreisfr commented:
@arsenm , may I ask you to review PR again, with latest fixes and improvements?

arsenm commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.td:
```diff
@@ -986,6 +986,64 @@ let Predicates = [HasDiv32] in {
+  def IDTLB : RRR_Inst<0x00, 0x00, 0x05, (outs), (ins AR:$s),
+                     "idtlb\t$s", []> {
+    let r = 0xC;
+	  let t = 0x0;
```
Weird indent

arsenm commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.td:
```diff
@@ -986,6 +986,64 @@ let Predicates = [HasDiv32] in {
+  def IITLB : RRR_Inst<0x00, 0x00, 0x05, (outs), (ins AR:$s),
+                     "iitlb\t$s", []> {
+    let r = 0x4;
+	  let t = 0x0;
```
Weird indent

MaskRay commented:
We ensure that llvm-project compiles without any warning with a recent Clang, but the latest the latest Xtensa target has some warnings.

I typically use this prebuilt Clang from chromium
 https://raw.githubusercontent.com/chromium/chromium/main/tools/clang/scripts/update.py
